### PR TITLE
Replace HTTPotion by HTTPoison

### PR DESCRIPTION
**Pull request by jmnsf**
This commit replaces the deprecated [HTTPotion](https://github.com/unrelentingtech/httpotion) by the hackney-based [HTTPoison](https://github.com/edgurgel/httpoison).

Beyond the fact that we're no longer dependent on a deprecated lib, this also enables connection pooling, which can make sequential requests to the Bot Framework API faster.

Since this is a major change, I've also included a bit of a refactor to the Client modules.

For the changelog
-----------------

**Replaced** `ExMicrosoftBot.Client.headers/2` by `ExMicrosoftBot.Client.authed_headers/1`, which removes the need for fetching the auth token to build valid request headers.

**Replaced** `ExMicrosoftBot.Client.req_opts/0` by `ExMicrosoftBot.Client.opts/1`.

**Removed** the `http_timeout` config and introduced the `http_client_opts` config. There doesn't seem to be an analog in HTTPoison for this as it looks like it was used by `:ibrowse` as a default value for [connection](https://github.com/cmullaparthi/ibrowse/blob/efd040de55f7a77d0b504e6c9de1849181bf64a6/src/ibrowse_http_client.erl#L800) and
[request](https://github.com/cmullaparthi/ibrowse/blob/efd040de55f7a77d0b504e6c9de1849181bf64a6/src/ibrowse_http_client.erl#L920) timeouts. Should be replaced by `http_client_opts: [timeout: X]` and `http_client_opts: [recv_timeout: X]` respectively.

**Adds** `ExMicrosoftBot.Client.get/2`, `ExMicrosoftBot.Client.delete/2`, `ExMicrosoftBot.Client.put/3`, and `ExMicrosoftBot.Client.post/3`.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -59,15 +59,17 @@ defmodule ExMicrosoftbot.Mixfile do
...
       {:poison, "~> 4.0"},
       {:jose, "~> 1.7"},
       {:timex, "~> 3.0"},
       {:tzdata, "~> 1.0"},
       {:inch_ex, "~> 2.0.0", only: :docs},
       {:dialyxir, "~> 0.3", only: [:dev]},
       {:ex_doc, "~> 0.19", only: [:dev]},
-      {:bypass, "~> 1.0", only: :test}
+      {:bypass, "~> 1.0", only: :test},
+      # Required by bypass, incompatible with OTP 22 since 2.8.0:
```
Comment (jmnsf) [inline mix.exs:71]: The lib supports Elixir from 1.8, which is [compatible only with OTP 20-21](https://hexdocs.pm/elixir/1.8.2/compatibility-and-deprecations.html#compatibility-between-elixir-and-erlang-otp).

File: lib/attachments_client.ex [hunk 2]
```diff
@@ -3,33 +3,40 @@ defmodule ExMicrosoftBot.Client.Attachments do
...
   alias ExMicrosoftBot.Models, as: Models
   alias ExMicrosoftBot.Client
 
   @doc """
-  Get AttachmentInfo structure describing the attachment views. [API Reference](https://docs.botframework.com/en-us/restapi/connector/#!/Attachments/Attachments_GetAttachmentInfo)
+  Get AttachmentInfo structure describing the attachment views.
+  @see [API Reference](https://docs.botframework.com/en-us/restapi/connector/#!/Attachments/Attachments_GetAttachmentInfo)
   """
-  @spec get_attachment(String.t, String.t) :: {:ok, Models.AttachmentInfo.t} | Client.error_type
+  @spec get_attachment(String.t(), String.t()) ::
```
Comment (zabirauf) [inline lib/attachments_client.ex]: Suggestion: You can improve the spec definition

```
@spec get_attachment(service_url :: String.t(), attachment_id :: String.t()) :: {:ok, Models.AttachmentInfo.t()} | Client.error_type()
```

File: lib/attachments_client.ex [hunk 3]
```diff
@@ -3,33 +3,40 @@ defmodule ExMicrosoftBot.Client.Attachments do
...
     |> deserialize_response(&Models.AttachmentInfo.parse/1)
   end
 
   @doc """
-  Get the named view as binary content. [API Reference](https://docs.botframework.com/en-us/restapi/connector/#!/Attachments/Attachments_GetAttachment)
+  Get the named view as binary content.
+  @see [API Reference](https://docs.botframework.com/en-us/restapi/connector/#!/Attachments/Attachments_GetAttachment)
   """
-  @spec get_attachment_view(String.t, String.t, String.t) :: {:ok, binary} | Client.error_type
+  @spec get_attachment_view(String.t(), String.t(), String.t()) ::
```
Comment (zabirauf) [inline lib/attachments_client.ex]: Suggestion: Similar here

**Review by zabirauf — APPROVED** [review]
Thanks for the changes, it looks good. Had small suggestions, otherwise it can be merged.

**Comment by jmnsf** [comment]
@zabirauf done! Thanks for the review.

File: lib/attachments_client.ex [hunk 4]
```diff
@@ -3,33 +3,40 @@ defmodule ExMicrosoftBot.Client.Attachments do
...
+          {:ok, String.t()} | Client.error_type()
```
Comment (zabirauf) [inline lib/attachments_client.ex]: @jmnsf the return type got changed from binary to String. Was that intentional as seems like the API still returns binary.

Comment (jmnsf) [reply lib/attachments_client.ex]: @zabirauf oversight, thanks!

**Comment by zabirauf** [comment]
@jmnsf FYI, published the new version to Hex. Thanks for the change

https://hex.pm/packages/ex_microsoftbot/3.0.0

**Pull request merged**